README.md
# guile-nacl
guile bindings for libsodium (nacl)


NEWS
Guile-nacl NEWS


Guile-nacl wraps the libsodium library for Guile Scheme.


guile-nacl 0.0.1 -- 2016-09-16
================================

* Initial release.


Copyright (C) 2016 Cryptotronix, LLC <jbd@cryptotronix.com>

Copying and distribution of this file, with or without modification, are
permitted in any medium without royalty provided the copyright notice
and this notice are preserved.


nacl.scm
;; guile-zmq
;; Copyright (C) 2011,2012 Andy Wingo <wingo at pobox dot com>

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU Lesser General Public License as
;; published by the Free Software Foundation; either version 3 of the
;; License, or (at your option) any later version.
;;
;; This program is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;; Lesser General Public License for more details.
;;
;; You should have received a copy of the GNU Lesser General Public
;; License along with this program; if not, contact:
;;
;; Free Software Foundation, Inc.     Voice:  +1-617-542-5942
;; 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652
;; Boston, MA  02110-1301,  USA       gnu@gnu.org

;;; Commentary:
;;
;; This is the libsodium wrapper for Guile.
;;
;; See the libsodium documentation for more details.
;;
;;; Code:

(define-module (nacl)
  #:use-module (nacl config)
  #:use-module (rnrs bytevectors)
  #:export (nacl-version

            nacl-init
            nacl-rand-buf))

;; This will export many things
(dynamic-call "scm_init_nacl" (dynamic-link *nacl-lib-path*))


#;
(if (not (member *zmq-documentation-path* documentation-files))
    (set! documentation-files (cons *zmq-documentation-path*
                                    documentation-files)))


guile-nacl/guile-nacl.h
/* guile-nacl
 * Copyright (C) 2016 Josh Datko <jbd@cryptotronix.com>
 *
 * guile-nacl.h:
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this program; if not, contact:
 *
 * Free Software Foundation, Inc.     Voice:  +1-617-542-5942
 * 51 Franklin St, Fifth Floor        Fax:    +1-617-542-2652
 * Boston, MA  02110-1301,  USA       gnu@gnu.org
 */

#ifndef __GUILE_NACL_H__
#define __GUILE_NACL_H__

#include <sodium.h>
#include <libguile.h>


void scm_init_nacl (void);


#endif


guile-nacl/guile-nacl.c
/* guile-zmq
 * Copyright (C) 2011, 2012 Andy Wingo <wingo@pobox.com>
 *
 * guile-zmq.c: 0MQ for Guile
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this program; if not, contact:
 *
 * Free Software Foundation, Inc.     Voice:  +1-617-542-5942
 * 51 Franklin St, Fifth Floor        Fax:    +1-617-542-2652
 * Boston, MA  02110-1301,  USA       gnu@gnu.org
 */

#include <libguile.h>

#include <sodium.h>
#include <assert.h>

#include "guile-nacl.h"


/* static scm_t_bits scm_tc16_zmq_context; */
/* static scm_t_bits scm_tc16_zmq_socket; */

/* /\* Map sockets to contexts, to keep contexts alive.  Also, allows atexit() to */
/*    close sockets and terminate contexts, as is done with ports.  *\/ */
/* static SCM all_sockets = SCM_BOOL_F; */
/* static scm_i_pthread_mutex_t all_sockets_lock = SCM_I_PTHREAD_MUTEX_INITIALIZER; */


/* static void */
/* register_socket (SCM sock, SCM ctx) */
/* { */
/*   scm_i_pthread_mutex_lock (&all_sockets_lock); */
/*   scm_hashq_set_x (all_sockets, sock, ctx); */
/*   scm_i_pthread_mutex_unlock (&all_sockets_lock); */
/* } */


/* static SCM scm_from_zmq_context (void *context) */
/* { */
/*   SCM_RETURN_NEWSMOB (scm_tc16_zmq_context, context); */
/* } */

/* static void* scm_to_zmq_context (SCM obj) */
/* { */
/*   void *ret; */

/*   SCM_ASSERT (SCM_SMOB_PREDICATE (scm_tc16_zmq_context, obj), */
/*               obj, 0, NULL); */
/*   ret = (void*)SCM_SMOB_DATA (obj); */

/*   if (!ret) */
/*     scm_misc_error (NULL, "terminated zmq context: ~a", scm_list_1 (obj)); */

/*   return ret; */
/* } */

/* static size_t scm_zmq_context_free (SCM obj) */
/* { */
/*   void *ctx = (void*)SCM_SMOB_DATA (obj); */
/*   if (ctx) */
/*     { fprintf (stderr, "terminating %p\n", ctx); */
/*       SCM_SET_SMOB_DATA (obj, 0); */
/*       zmq_term (ctx); */
/*     } */
/*   return 0; */
/* } */


/* static SCM scm_from_zmq_socket (void *socket) */
/* { */
/*   SCM_RETURN_NEWSMOB (scm_tc16_zmq_socket, socket); */
/* } */

/* static void* scm_to_zmq_socket (SCM obj) */
/* { */
/*   void *ret; */

/*   SCM_ASSERT (SCM_SMOB_PREDICATE (scm_tc16_zmq_socket, obj), */
/*               obj, 0, NULL); */
/*   ret = (void*)SCM_SMOB_DATA (obj); */

/*   if (!ret) */
/*     scm_misc_error (NULL, "closed zmq socket: ~a", scm_list_1 (obj)); */

/*   return ret; */
/* } */

/* static size_t scm_zmq_socket_free (SCM obj) */
/* { */
/*   void *sock = (void*)SCM_SMOB_DATA (obj); */

/*   if (sock) */
/*     { */
/*       zmq_close (sock); */
/*       SCM_SET_SMOB_DATA (obj, 0); */
/*     } */

/*   return 0; */
/* } */


/* static void */
/* scm_zmq_error (const char *subr) */
/* { */
/*   int e = errno; */

/*   scm_error (scm_from_locale_symbol ("zmq-error"), */
/*              subr, */
/*              zmq_strerror (e), */
/*              SCM_EOL, */
/*              scm_list_1 (scm_from_int (e))); */
/* } */


/* SCM_DEFINE (scm_zmq_version, "zmq-version", 0, 0, 0, */
/* 	    (void), */
/* 	    "Retrieves the version of the zmq library.") */
/* #define FUNC_NAME s_scm_zmq_version */
/* { */
/*   int major = 0, minor = 0, patch = 0; */
/*   zmq_version (&major, &minor, &patch); */
/*   return scm_values (scm_list_3 (scm_from_int (major), */
/*                                  scm_from_int (minor), */
/*                                  scm_from_int (patch))); */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_init, "zmq-init", 0, 1, 0, */
/*             (SCM io_threads), */
/*             "Create a ZeroMQ context with @var{io_threads} threads.") */
/* #define FUNC_NAME s_scm_zmq_init */
/* { */
/*   int threads; */
/*   void *ctx; */

/*   if (SCM_UNBNDP (io_threads)) */
/*     threads = 1; */
/*   else */
/*     threads = scm_to_int (io_threads); */

/*   ctx = zmq_init (threads); */
/*   if (!ctx) */
/*     scm_zmq_error (FUNC_NAME); */

/*   return scm_from_zmq_context (ctx); */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_term, "zmq-term", 1, 0, 0, */
/*             (SCM context), */
/*             "Terminate the ZeroMQ context @var{context}.") */
/* #define FUNC_NAME s_scm_zmq_term */
/* { */
/*   void *ctx = scm_to_zmq_context (context); */

/*   if (zmq_term (ctx)) */
/*     scm_zmq_error (FUNC_NAME); */
/*   SCM_SET_SMOB_DATA (context, NULL); */

/*   return SCM_UNSPECIFIED; */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_socket, "zmq-socket", 2, 0, 0, */
/* 	    (SCM context, SCM type), */
/* 	    "Create a new ZeroMQ of type @var{type} within context @var{context}.") */
/* #define FUNC_NAME s_scm_zmq_socket */
/* { */
/*   void *ctx, *s; */
/*   int typ; */
/*   SCM ret; */

/*   ctx = scm_to_zmq_context (context); */
/*   typ = scm_to_int (type); */

/*   s = zmq_socket (ctx, typ); */
/*   if (!s) */
/*     scm_zmq_error (FUNC_NAME); */

/*   ret = scm_from_zmq_socket (s); */
/*   register_socket (ret, context); */

/*   return ret; */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_close, "zmq-close", 1, 0, 0, */
/* 	    (SCM socket), */
/* 	    "Close @var{socket}.") */
/* #define FUNC_NAME s_scm_zmq_close */
/* { */
/*   void *s; */

/*   s = scm_to_zmq_socket (socket); */

/*   if (zmq_close (s)) */
/*     scm_zmq_error (FUNC_NAME); */
/*   SCM_SET_SMOB_DATA (socket, NULL); */

/*   return SCM_UNSPECIFIED; */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_setsockopt, "%zmq-setsockopt", 3, 0, 0, */
/* 	    (SCM socket, SCM opt, SCM val), */
/* 	    "Set @var{opt} to @var{val} on @var{socket}.") */
/* #define FUNC_NAME s_scm_zmq_setsockopt */
/* { */
/*   void *s; */
/*   int o; */

/*   s = scm_to_zmq_socket (socket); */
/*   o = scm_to_int (opt); */
/*   SCM_VALIDATE_BYTEVECTOR (3, val); */

/*   if (zmq_setsockopt (s, o, SCM_BYTEVECTOR_CONTENTS (val), */
/*                       SCM_BYTEVECTOR_LENGTH (val))) */
/*     scm_zmq_error (FUNC_NAME); */

/*   return SCM_UNSPECIFIED; */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_getsockopt, "%zmq-getsockopt", 3, 0, 0, */
/* 	    (SCM socket, SCM opt, SCM val), */
/* 	    "Get @var{opt} of @var{socket} into @var{val}.") */
/* #define FUNC_NAME s_scm_zmq_getsockopt */
/* { */
/*   void *s; */
/*   int o; */
/*   size_t len; */

/*   s = scm_to_zmq_socket (socket); */
/*   o = scm_to_int (opt); */
/*   SCM_VALIDATE_BYTEVECTOR (3, val); */
/*   len = SCM_BYTEVECTOR_LENGTH (val); */

/*   if (zmq_getsockopt (s, o, SCM_BYTEVECTOR_CONTENTS (val), &len)) */
/*     scm_zmq_error (FUNC_NAME); */

/*   if (len != SCM_BYTEVECTOR_LENGTH (val)) */
/*     scm_misc_error (FUNC_NAME, "bad length of socket option: ~a", */
/*                     scm_from_size_t (len)); */

/*   return SCM_UNSPECIFIED; */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_bind, "zmq-bind", 2, 0, 0, */
/* 	    (SCM socket, SCM addr), */
/* 	    "Bind @var{socket} to @var{addr}.") */
/* #define FUNC_NAME s_scm_zmq_bind */
/* { */
/*   void *s; */
/*   char *caddr; */

/*   scm_dynwind_begin (0); */

/*   s = scm_to_zmq_socket (socket); */
/*   caddr = scm_to_locale_string (addr); */

/*   scm_dynwind_free (caddr); */

/*   if (zmq_bind (s, caddr)) */
/*     scm_zmq_error (FUNC_NAME); */

/*   scm_dynwind_end (); */

/*   return SCM_UNSPECIFIED; */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_connect, "zmq-connect", 2, 0, 0, */
/* 	    (SCM socket, SCM addr), */
/* 	    "Connect to @var{addr} on @var{socket}.") */
/* #define FUNC_NAME s_scm_zmq_connect */
/* { */
/*   void *s; */
/*   char *caddr; */

/*   scm_dynwind_begin (0); */

/*   s = scm_to_zmq_socket (socket); */
/*   caddr = scm_to_locale_string (addr); */

/*   scm_dynwind_free (caddr); */

/*   if (zmq_connect (s, caddr)) */
/*     scm_zmq_error (FUNC_NAME); */

/*   scm_dynwind_end (); */

/*   return SCM_UNSPECIFIED; */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_send, "zmq-send", 2, 1, 0, */
/* 	    (SCM socket, SCM bv, SCM flags), */
/* 	    "Send @var{bv} over @var{socket}.") */
/* #define FUNC_NAME s_scm_zmq_send */
/* { */
/*   void *s; */
/*   int cflags; */
/*   zmq_msg_t msg; */

/*   s = scm_to_zmq_socket (socket); */
/*   SCM_VALIDATE_BYTEVECTOR (2, bv); */
/*   if (SCM_UNBNDP (flags)) */
/*     cflags = 0; */
/*   else */
/*     cflags = scm_to_int (flags); */

/*   if (zmq_msg_init_size (&msg, SCM_BYTEVECTOR_LENGTH (bv))) */
/*     scm_zmq_error (FUNC_NAME); */

/*   memcpy (zmq_msg_data (&msg), SCM_BYTEVECTOR_CONTENTS (bv), */
/*           SCM_BYTEVECTOR_LENGTH (bv)); */

/*   if (zmq_send (s, &msg, cflags)) */
/*     { */
/*       int errno_save = errno; */
/*       zmq_msg_close (&msg); */
/*       errno = errno_save; */
/*       scm_zmq_error (FUNC_NAME); */
/*     } */

/*   zmq_msg_close (&msg); */

/*   return SCM_UNSPECIFIED; */
/* } */
/* #undef FUNC_NAME */


/* SCM_DEFINE (scm_zmq_recv, "zmq-recv", 1, 1, 0, */
/* 	    (SCM socket, SCM flags), */
/* 	    "Receive a message from @var{socket}.") */
/* #define FUNC_NAME s_scm_zmq_recv */
/* {p */

SCM_DEFINE (scm_nacl_rand_buf, "nacl-rand-buf", 1, 0, 0,
	    (SCM len),
	    "Returns a random buffer of length @var{len}.")
#define FUNC_NAME s_scm_nacl_rand_buf
{
    SCM buf = scm_c_make_bytevector (scm_to_size_t (len));
    randombytes_buf (SCM_BYTEVECTOR_CONTENTS (buf), scm_to_size_t (len));

    return buf;
}
#undef FUNC_NAME



void
scm_init_nacl (void)
{
    static int initialized = 0;

    if (initialized)
        return;

    int rc = sodium_init();
    assert (-1 != rc);

#ifndef SCM_MAGIC_SNARFER
#include "guile-nacl.x"
#endif


    initialized = 1;
}
